MessageFormatter parse error improvements

pilif opened the pull request:
Looking into `MessageFormatter` and the underlying ICU pattern documentation, I was blown away by the functionality offered by the patterns that `MessageFormatter` understands. This is incredibly useful and I was totally going to look into making use of this in our main product which already relies on `intl` for number and date formatting, but uses a home-grown solution for message localization.

However, while playing around with this, I came across a usability-issue around `MessageFormatter::formatMessage()` and especially `MessageFormatter::__construct()`:

As the formatting patterns can be pretty complex, it's likely that they contain parsing errors (especially in my case as I was *just* getting into this), so it would be helpful to  get to details of such parsing errors.

Unfortunately, `MessageFormatter::formatMessage()` via `intl_get_error_message()` would just report a generic `U_PATTERN_SYNTAX_ERROR` and, even worse, `MessageFormatter:__construct()` would turn *any* possible internal error into a generic `Constructor failed` error message.

It doesn't have to be that way though: ICU is capable of reporting additional information about a parser error and indeed, the `intl` extension is already making use of this in other places.

So this PR contains two commits that

1. fix the problem for `MessageFormatter::formatMessage()` and `intl_get_error_message()` to report the additional parsing error message (see updated tests)
2. fix `MessageFormatter::__construct()` to report the actual error message rather than a generic "Constructor failed"

pilif commented:
oh crap - you've *just* cut the PHP-8 branch.

If you think this could go into PHP 8.0, I'll rebase and re-target the PR

Girgias commented:
Test failure is unrelated and has been marked as XFAIL on master.

nikic commented on ext/intl/msgformat/msgformat.c:
```diff
@@ -74,12 +76,31 @@ static int msgfmt_ctor(INTERNAL_FUNCTION_PARAMETERS)
+		smart_str_free( &parse_error_str );
+		intl_error_set_code( NULL, INTL_DATA_ERROR_CODE( mfo ) );
+
+		if( msg != NULL )
```
This null check is unnecessary, spprintf always produces a non-NULL string.

pilif replied:
I'll add a mixup (and rebase on top of 7.4), but FYI, this code is somewhat "heavily inspired" by https://github.com/php/php-src/blob/c3365bb301fd0a0cb6b5e3b6df5f6742d4262dca/ext/intl/transliterator/transliterator_methods.c#L171-L179

I'll make a followup PR to get rid of the null check there too 😝

nikic commented on ext/intl/msgformat/msgformat_format.c:
```diff
@@ -119,10 +120,30 @@ PHP_FUNCTION( msgfmt_format_message )
+		smart_str_free( &parse_error_str );
+		intl_error_set_code( NULL, INTL_DATA_ERROR_CODE( mfo ) );
+
+		if( msg != NULL )
```
Again, this check is not needed.

nikic reviewed (commented):
This looks good to me, and I think it should go into PHP 7.4 (unless it's conflict-y). Doesn't change any behavior, just improves diagnostics.

pilif commented:
there will be a few conflicts on PHP-7.4 (mostly due to format string changes in tests). I'll make a second PR targeting 7.4 - then you can pick which one you'd be willing to merge.

pilif commented:
rebased on top of current `master` (which means that the false-positive failure due to the `timelib` changes now went away) and implemented the suggestion by @nikic.

I have also made a secondary PR (#6325) on top of `PHP-7.4` if you prefer to merge this into 7.4

nikic commented:
Closing this as I merged #6325.
